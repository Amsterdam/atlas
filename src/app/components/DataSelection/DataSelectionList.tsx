import { Link, themeSpacing } from '@amsterdam/asc-ui'
import { FunctionComponent } from 'react'
import { Link as RouterLink } from 'react-router-dom'
import styled from 'styled-components'
import { getDetailPageData } from '../../../store/redux-first-router/actions'
import buildDetailUrl from '../../pages/MapPage/detail/buildDetailUrl'
import DataSelectionFormatter from './DataSelectionFormatter/DataSelectionFormatter'
import { Data } from './types'

const StyledListItem = styled.li`
  margin-bottom: ${themeSpacing(2)};
`

const DataSelectionList: FunctionComponent<{
  content: Data
}> = ({ content }) => (
  <ul data-testid="dataSelectionList">
    {content.body.map((row) => (
      <StyledListItem key={row.id}>
        <Link as={RouterLink} to={buildDetailUrl(getDetailPageData(row.detailEndpoint))} inList>
          {/*
          // @ts-ignore */}
          <DataSelectionFormatter
            // @ts-ignore
            variables={row.content[0]}
            formatter={content.formatters[0]}
            useInline
          />
        </Link>

        {row.content.map(
          (variables, i) =>
            i !== 0 && (
              // @ts-ignore
              <DataSelectionFormatter
                // eslint-disable-next-line react/no-array-index-key
                key={i}
                // @ts-ignore
                variables={variables}
                formatter={content.formatters[i]}
                useInline
              />
            ),
        )}
      </StyledListItem>
    ))}
  </ul>
)

export default DataSelectionList
