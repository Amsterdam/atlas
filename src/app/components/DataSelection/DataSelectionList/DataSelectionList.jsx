/* eslint-disable react/no-array-index-key */
import React from 'react';
import PropTypes from 'prop-types';
import { connect } from 'react-redux';
import { bindActionCreators } from 'redux';
import { toDetailFromEndpoint } from '../../../../store/redux-first-router/actions';
import DataSelectionFormatter from '../DataSelectionFormatter/DataSelectionFormatter';

const DataSelectionList = ({ content, navigateToDetail }) => (
  <ul className="o-list u-margin__bottom--1">
    {content.body.map((row, index) => (
      <li key={`${index}/${row.detailEndpoint}`}>
        <button
          className="o-btn o-btn--link"
          onClick={() => navigateToDetail(row.detailEndpoint)}
        >
          <DataSelectionFormatter
            variables={row.content[0]}
            formatter={content.formatters[0]}
            useInline
          />
        </button>

        {row.content.map((variables, i) => (
          (i !== 0) && <DataSelectionFormatter
            key={i}
            variables={variables}
            formatter={content.formatters[i]}
            useInline
          />
        ))}
      </li>
    ))}
  </ul>
);

DataSelectionList.propTypes = {
  content: PropTypes.shape({
    head: PropTypes.array,
    body: PropTypes.array
  }).isRequired,
  navigateToDetail: PropTypes.func.isRequired
};


const mapDispatchToProps = (dispatch) => bindActionCreators({
  navigateToDetail: toDetailFromEndpoint
}, dispatch);

export default connect(null, mapDispatchToProps)(DataSelectionList);
