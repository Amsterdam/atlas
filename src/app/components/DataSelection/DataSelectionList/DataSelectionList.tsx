import { FunctionComponent } from 'react'
import { generatePath, Link as RouterLink } from 'react-router-dom'
import styled from 'styled-components'
import { Link, themeSpacing } from '@amsterdam/asc-ui'
import { LocationDescriptor } from 'history'
import { getDetailPageData } from '../../../../store/redux-first-router/actions'
import DataSelectionFormatter from '../DataSelectionFormatter/DataSelectionFormatter'
import { routing } from '../../../routes'

const StyledListItem = styled.li`
  margin-bottom: ${themeSpacing(2)};
`

export const toDetailPage = (endpoint: string): LocationDescriptor => {
  const { type, subtype, id } = getDetailPageData(endpoint)
  return {
    pathname: generatePath(routing.dataDetail.path, { type, subtype, id }),
  }
}

// TODO: Accurately determine the types in this interface.
export interface DataSelectionListProps {
  content: {
    head: []
    body: [
      {
        detailEndpoint: string
        content: string[]
      },
    ]
    formatters: string[]
  }
}

const DataSelectionList: FunctionComponent<DataSelectionListProps> = ({ content }) => {
  return (
    <ul>
      {content.body.map((row) => (
        <StyledListItem key={row.content[0]}>
          <Link as={RouterLink} to={toDetailPage(row.detailEndpoint)} inList>
            {/*
            // @ts-ignore */}
            <DataSelectionFormatter
              // @ts-ignore
              variables={row.content[0]}
              formatter={content.formatters[0]}
              useInline
            />
          </Link>

          {row.content.map(
            (variables, i) =>
              i !== 0 && (
                // @ts-ignore
                <DataSelectionFormatter
                  // eslint-disable-next-line react/no-array-index-key
                  key={i}
                  // @ts-ignore
                  variables={variables}
                  formatter={content.formatters[i]}
                  useInline
                />
              ),
          )}
        </StyledListItem>
      ))}
    </ul>
  )
}

export default DataSelectionList
