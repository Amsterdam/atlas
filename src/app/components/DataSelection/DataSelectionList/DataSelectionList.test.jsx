import React from 'react'
import configureMockStore from 'redux-mock-store'
import { shallow } from 'enzyme'
import DataSelectionList from './DataSelectionList'

jest.mock('react-redux')

describe('DataSelectionList', () => {
  it('should render without failing', () => {
    const store = configureMockStore()({})
    const component = shallow(
      <DataSelectionList
        store={store}
        content={{
          body: [
            {
              detailEndpoint: 'http://example.com',
              content: [[{}]],
            },
          ],
          formatters: [],
        }}
        navigateToDetail={() => {}}
      />,
    )

    expect(component).toMatchSnapshot()
  })
})
