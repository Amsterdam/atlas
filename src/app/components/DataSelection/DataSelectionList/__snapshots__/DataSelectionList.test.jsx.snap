// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`DataSelectionList should render without failing 1`] = `
<ul
  className="o-list u-margin__bottom--1"
>
  <li
    key="0/http://example.com"
  >
    <button
      className="o-btn o-btn--link"
      onClick={[Function]}
      type="button"
    >
      <DataSelectionFormatter
        useInline={true}
        variables={
          Array [
            Object {},
          ]
        }
      />
    </button>
  </li>
</ul>
`;
